In [1]:
# Task 1

import pandas as pd

In [2]:
df = pd.read_csv(r'C:\KATE\survey_results_public.csv',
               low_memory=False)

In [3]:
df.head()

,ResponseId,MainBranch,Age,EdLevel,Employment,EmploymentAddl,WorkExp,LearnCodeChoose,LearnCode,LearnCodeAI,...,AIAgentOrchestration,AIAgentOrchWrite,AIAgentObserveSecure,AIAgentObsWrite,AIAgentExternal,AIAgentExtWrite,AIHuman,AIOpen,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,25-34 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Employed,"Caring for dependents (children, elderly, etc.)",8.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,Vertex AI,NaN,NaN,NaN,ChatGPT,NaN,When I don’t trust AI’s answers,"Troubleshooting, profiling, debugging",61256.0,10.0
1,2,I am a developer by profession,25-34 years old,"Associate degree (A.A., A.S., etc.)",Employed,NaN,2.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers;When I want to...,All skills. AI is a flop.,104413.0,9.0
2,3,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)","Independent contractor, freelancer, or self-em...",None of the above,10.0,"Yes, I am not new to coding but am learning ne...",Online Courses or Certification (includes all ...,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code;GitHub Copilot;Google Gemini,NaN,When I don’t trust AI’s answers;When I want to...,"Understand how things actually work, problem s...",53061.0,8.0
3,4,I am a developer by profession,35-44 years old,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Employed,None of the above,4.0,"Yes, I am not new to coding but am learning ne...","Other online resources (e.g. standard search, ...","Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,ChatGPT;Claude Code,NaN,When I don’t trust AI’s answers;When I want to...,NaN,36197.0,6.0
4,5,I am a developer by profession,35-44 years old,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)","Independent contractor, freelancer, or self-em...","Caring for dependents (children, elderly, etc.)",21.0,"No, I am not new to coding and did not learn n...",NaN,"Yes, I learned how to use AI-enabled tools for...",...,NaN,NaN,NaN,NaN,NaN,NaN,When I don’t trust AI’s answers,"critical thinking, the skill to define the tas...",60000.0,7.0


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49191 entries, 0 to 49190
Columns: 172 entries, ResponseId to JobSat
dtypes: float64(52), int64(1), object(119)
memory usage: 64.6+ MB


In [5]:
total_respondents = len(df)

print(f'Загальна кількість респондентів: {total_respondents}')

Загальна кількість респондентів: 49191


In [6]:
# Task 2

import pandas as pd

df = pd.read_csv(r'C:\KATE\survey_results_public.csv',
               low_memory=False)

schema = pd.read_csv(r'C:\KATE\survey_results_schema.csv')

In [7]:
schema_questions = set(schema['qname'])

print(f'Питань у схемі: {len(schema_questions)}')

Питань у схемі: 139


In [8]:
df_columns = set(df.columns)

common_cols = schema_questions.intersection(df_columns)

print(f'Колонок у датасеті:   {len(df_columns)}')
print(f'Питань у схемі:       {len(schema_questions)}')
print(f'Спільних колонок:     {len(common_cols)}')

Колонок у датасеті:   172
Питань у схемі:       139
Спільних колонок:     126


In [9]:
complete_df = df[list(common_cols)].dropna()

complete_respondents = len(complete_df)

print(f'Респондентів без пропусків: {complete_respondents}')

Респондентів без пропусків: 0


Це очікуваний результат, оскільки:
- Опитування містить велику кількість питань (126)
- Багато з них є необов'язковими
- У реальних даних респонденти часто пропускають частину питань
- Навіть найактивніший респондент має щонайменше 4 пропущених відповіді.

In [10]:
# Task 3
df['WorkExp'].isna().sum()

# Розрахунок статистик
mean_exp = df['WorkExp'].dropna().mean()
median_exp = df['WorkExp'].median()
mode_exp = df['WorkExp'].mode()[0]

# Виведення результатів
print(f'Середній досвід (mean): {mean_exp:.2f}')
print(f'Медіана (median): {median_exp}')
print(f'Мода (mode): {mode_exp}')

Середній досвід (mean): 13.37
Медіана (median): 10.0
Мода (mode): 10.0


 Висновок:
- Середній досвід (13.37 років) є вищим за медіану (10 років),
що свідчить про наявність респондентів з дуже великим досвідом (викидів)
- Медіана та мода співпадають (10 років), отже це найбільш типовий досвід
- Більшість респондентів мають близько 10 років досвіду роботи

In [11]:
# Task 4

df['RemoteWork'].value_counts()

RemoteWork
Remote                                                                          10931
Hybrid (some remote, leans heavy to in-person)                                   6732
In-person                                                                        6042
Hybrid (some in-person, leans heavy to flexibility)                              5831
Your choice (very flexible, you can come in when you want or just as needed)     4244
Name: count, dtype: int64

In [12]:
# Фільтрація тільки remote
remote_workers = df[df['RemoteWork'] == 'Remote']

# Кількість
remote_count = len(remote_workers)

print(f'Кількість респондентів, які працюють віддалено: {remote_count}')

Кількість респондентів, які працюють віддалено: 10931


In [13]:
# Task 5

df.columns

Index(['ResponseId', 'MainBranch', 'Age', 'EdLevel', 'Employment',
       'EmploymentAddl', 'WorkExp', 'LearnCodeChoose', 'LearnCode',
       'LearnCodeAI',
       ...
       'AIAgentOrchestration', 'AIAgentOrchWrite', 'AIAgentObserveSecure',
       'AIAgentObsWrite', 'AIAgentExternal', 'AIAgentExtWrite', 'AIHuman',
       'AIOpen', 'ConvertedCompYearly', 'JobSat'],
      dtype='object', length=172)

In [14]:
df['LanguageHaveWorkedWith'].head()

0                     Bash/Shell (all shells);Dart;SQL
1                                                 Java
2                  Dart;HTML/CSS;JavaScript;TypeScript
3                                      Java;Kotlin;SQL
4    C;C#;C++;Delphi;HTML/CSS;Java;JavaScript;Lua;P...
Name: LanguageHaveWorkedWith, dtype: object

In [15]:
python_users = df['LanguageHaveWorkedWith'].str.contains('Python', na=False)

python_count = python_users.sum()
total = len(df)

python_percentage = (python_count / total) * 100

print(f'Відсоток респондентів, які використовують Python: {python_percentage:.2f}%')

Відсоток респондентів, які використовують Python: 37.54%


Висновок:
37.54% респондентів використовують Python у своїй роботі.
Це означає, що Python є однією з найпопулярніших мов програмування серед розробників.
Оскільки респонденти могли обирати кілька мов, цей відсоток не є взаємовиключним
і може перетинатися з іншими мовами (наприклад, JavaScript, SQL тощо).

In [16]:
# Task 6

df['LearnCode'].head()

0    Online Courses or Certification (includes all ...
1    Online Courses or Certification (includes all ...
2    Online Courses or Certification (includes all ...
3    Other online resources (e.g. standard search, ...
4                                                  NaN
Name: LearnCode, dtype: object

In [17]:
online_learners = df['LearnCode'].str.contains('Online Courses', na=False)

online_count = online_learners.sum()

print(f'Кількість респондентів, які навчались через онлайн курси: {online_count}')

Кількість респондентів, які навчались через онлайн курси: 10973


Висновок:
10973 респонденти зазначили, що використовували онлайн курси для навчання програмуванню.
Це свідчить про високу популярність онлайн-освіти серед розробників.
Оскільки респонденти могли обирати кілька варіантів навчання,
онлайн курси часто використовуються разом з іншими джерелами (книги, практика, університет тощо).

In [18]:
# Task 7

# 1. Фільтр Python-розробників
python_df = df[df['LanguageHaveWorkedWith'].str.contains('Python', na=False)]

# 2. Прибираємо пропуски в компенсації
python_df = python_df.dropna(subset=['ConvertedCompYearly', 'Country'])

# 3. Групування по країнах + агрегація
salary_by_country = python_df.groupby('Country')['ConvertedCompYearly'].agg(['mean', 'median'])

# 4. Перейменування колонок (для краси)
salary_by_country = salary_by_country.rename(columns={
    'mean': 'AverageSalary',
    'median': 'MedianSalary'
})

# 5. Сортування (опціонально, але красиво)
salary_by_country = salary_by_country.sort_values(by='AverageSalary', ascending=False)

# Вивід
salary_by_country.head(10)

,AverageSalary,MedianSalary
Country,,
Oman,390135.000000,390135.0
Andorra,226103.500000,226103.5
Viet Nam,218837.166667,8254.0
United States of America,173298.590211,150000.0
Switzerland,156456.600000,142592.0
Singapore,147515.121951,91391.0
Israel,135828.365385,142594.0
Nigeria,128773.276596,1808.0
Ireland,120523.918919,116015.0


Висновок:
Рівень компенсації Python-розробників суттєво відрізняється між країнами.
У деяких країнах спостерігається велика різниця між середнім і медіанним значенням,
що свідчить про наявність викидів (дуже високих або дуже низьких зарплат).
Наприклад, у Viet Nam та Nigeria середнє значення значно перевищує медіану,
що означає, що невелика кількість респондентів з високими доходами впливає на середнє.
Медіана є більш надійним показником "типового" доходу.

In [19]:
# Task 8

# Прибираємо пропуски в зарплаті
salary_df = df.dropna(subset=['ConvertedCompYearly'])

# Сортуємо за спаданням зарплати
top_paid = salary_df.sort_values(by='ConvertedCompYearly', ascending=False)

# Беремо топ-5 і дивимось освіту
top5_education = top_paid[['ConvertedCompYearly', 'EdLevel']].head(5)

print(top5_education)

       ConvertedCompYearly                                          EdLevel
34267           50000000.0              Associate degree (A.A., A.S., etc.)
28700           33552715.0  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)
43143           18387548.0              Associate degree (A.A., A.S., etc.)
35353           15430267.0     Bachelor’s degree (B.A., B.S., B.Eng., etc.)
45971           13921760.0  Master’s degree (M.A., M.S., M.Eng., MBA, etc.)


Висновок:
Найбільш оплачувані респонденти мають різні рівні освіти,
включаючи Associate, Bachelor’s та Master’s degree.
Це свідчить про те, що високий рівень доходу не обов'язково
прямо залежить від формальної освіти.

In [20]:
# Task 9

df['UsesPython'] = df['LanguageHaveWorkedWith'].str.contains('Python', na=False)

# Групування по віку
age_stats = df.groupby('Age').agg(
    total_respondents=('ResponseId', 'count'),
    python_users=('UsesPython', 'sum')
)

# Рахуємо відсоток
age_stats['PythonPercentage'] = (age_stats['python_users'] / age_stats['total_respondents'] * 100).round(2)

result = age_stats.sort_values('PythonPercentage', ascending=False)[['total_respondents', 'python_users', 'PythonPercentage']]
result['PythonPercentage'] = result['PythonPercentage'].astype(str) + '%'

print(result)

                   total_respondents  python_users PythonPercentage
Age                                                                
18-24 years old                 9210          3684            40.0%
45-54 years old                 6275          2424           38.63%
55-64 years old                 2626           978           37.24%
25-34 years old                16519          6102           36.94%
35-44 years old                13241          4862           36.72%
65 years or older                942           298           31.63%
Prefer not to say                378           118           31.22%


Висновок:
Python найпопулярніший серед наймолодших розробників (18–24 роки) — 40%. У всіх інших вікових групах показник стабільно тримається на рівні 36–39%, і лише серед респондентів 65+ трохи нижчий — 31.63%. Це свідчить про те, що Python є популярним інструментом незалежно від віку.

In [21]:
# Task 10

# 1. Прибираємо пропуски в ключових колонках
clean_df = df.dropna(subset=['ConvertedCompYearly', 'RemoteWork', 'Industry'])

# 2. Обчислюємо 75-й перцентиль
q75 = clean_df['ConvertedCompYearly'].quantile(0.75)

# 3. Фільтруємо:
# - висока зарплата
# - remote робота
top_remote = clean_df[
    (clean_df['ConvertedCompYearly'] > q75) &
    (clean_df['RemoteWork'] == 'Remote')
]

# 4. Найпопулярніші індустрії
top_industries = top_remote['Industry'].value_counts()

print(top_industries.head(10))

Industry
Software Development                          1174
Fintech                                        188
Healthcare                                     186
Other:                                         176
Internet, Telecomm or Information Services     137
Banking/Financial Services                      87
Government                                      78
Media & Advertising Services                    75
Retail and Consumer Services                    65
Transportation, or Supply Chain                 63
Name: count, dtype: int64


Висновок:
Найбільш поширеною індустрією серед високооплачуваних віддалених працівників
є Software Development, що значно випереджає інші галузі.
Також популярними є Fintech, Healthcare та Internet/IT-сектор.
Це свідчить про те, що технологічні індустрії пропонують найкращі можливості
для віддаленої роботи з високим рівнем компенсації.

Загальний висновок:
У ході аналізу даних опитування Stack Overflow було досліджено
різні аспекти діяльності розробників, включаючи досвід роботи,
рівень компенсації, популярність мов програмування та формати роботи.
Python є однією з найпопулярніших мов серед розробників різного віку.
Найвищі зарплати спостерігаються у технологічних індустріях,
особливо серед віддалених працівників.
Дані містять викиди, тому при аналізі важливо використовувати медіану
разом із середнім значенням.